api/src/main/java/io/minio/MinioClient.java

```diff
@@ -2776,27 +2998,333 @@ public void putObject(String bucketName, String objectName, InputStream stream,
+
+    String keyEncryptionCipherMode = "RSA";
+    if (keypair == null) {
+      throw new InvalidArgumentException("empty key pair not allowed");
+    }
+    
+    putObject(bucketName, objectName, stream, size, contentType, keypair, keyEncryptionCipherMode);
+  }
+
+  /*
+   * Common method to encrypt data, encrypt keys and upload it to the server.
+   */
+  private void putObject(String bucketName, String objectName, InputStream stream, long size, String contentType,
+      Object key, String keyEncryptionCipherMode) throws NoSuchAlgorithmException, InvalidKeyException,
+      NoSuchPaddingException, IllegalBlockSizeException, BadPaddingException, InvalidAlgorithmParameterException,
+      IOException, InvalidArgumentException, InvalidBucketNameException, InsufficientDataException, NoResponseException,
+      ErrorResponseException, InternalException, XmlPullParserException {
+
+    String dataEncryptionCipherMode = "AES/CBC/PKCS5Padding";
+    Key masterKey = null;
+
+    // Generate symmetric 256 bit AES key for data encryption
+    KeyGenerator symKeyGenerator = KeyGenerator.getInstance("AES");
+    symKeyGenerator.init(256);
+    SecretKey dataEncryptionKey = symKeyGenerator.generateKey();
+
+    // Generate an iv to be used for data encryption
+    byte[] iv = new byte[16];
+    SecureRandom ivSeed = new SecureRandom();
+    ivSeed.nextBytes(iv);
+
+    // Get CipherStream after encryption of input stream
+    CipherInputStream cipherInputStream = Crypto.encryptInputStream(stream, dataEncryptionKey, dataEncryptionCipherMode,
+        iv);
+
+    // get the master key type and typecast
+    if (key instanceof KeyPair) {
+      masterKey = ((KeyPair) key).getPublic();
+    } else if (key instanceof SecretKey) {
+      masterKey = (SecretKey) key;
+    } else {
+      throw new InvalidArgumentException("key should be an instance of KeyPair or SecretKey");
+    }
+
+    // encrypt the plain data key using master key
+    byte[] dataKey = dataEncryptionKey.getEncoded();
+    byte[] encryptedDataKey = Crypto.encrypt(dataKey, masterKey, keyEncryptionCipherMode);
+
+    // Prepare data to be put to the object header
+    String encDataKey = BaseEncoding.base64().encode(encryptedDataKey);
+    String ivString = BaseEncoding.base64().encode(iv);
+    
+    // Set the headermap with encryption related metadata
+    Map<String, String> headerMap = new HashMap<>();
+    headerMap.put("x-amz-meta-x-amz-key", encDataKey);
+    headerMap.put("x-amz-meta-x-amz-iv", ivString);
+    // matdesc is unused
+    headerMap.put("x-amz-meta-x-amz-matdesc", "{}");
+
+    // Create byte array stream to hold encrypted stream
+    ByteArrayOutputStream encryptedStream = new ByteArrayOutputStream();
+
+    // Read bytes from cipher stream
+    int numRead = 0;
+    byte[] readBytes = new byte[8192];
+    while ((numRead = cipherInputStream.read(readBytes)) != -1) {
+      // write to byte array stream
+      encryptedStream.write(readBytes, 0, numRead);
+    }
```

Comment:
I have few questions here
* Why do we need intermittent `ByteArrayOutputStream/ByteArrayInputStream` as `cipherInputStream` is already an `InputStream`?
* We encrypt each 8192 bytes from the input stream (note that `numRead` might not be guaranteed to have 8192) and stores to object.  What will happen if `getObject()` reads less than 8192 bytes or higher than 8192 bytes to decrypt?
* Does encryption increase/decrease total size of input stream? In short, does `encryptedStream` contains same amount of data as passed input stream?
* Assume the input stream is having GiBs of data.  This code holds same amount of memory for `encryptedStream`, we should use `PipedInputStream/PipedOutputStream`  or our own input/output stream to handle encryption on the fly.